client/Contexts/Introspect/introspection.ts

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+import {
+  IntrospectionSchema,
+  IntrospectionType,
+  IntrospectionTypeRef,
+} from 'graphql';
+import {
+  Entities,
+  Entity,
+  Expectations,
+  OperationCallback,
+} from './ExpectationTypes';
+import fromEntries from 'fromentries';
+
+/**
+ * Transforms a field from an object representation to a GraphQL schema string representation
+ * @param type the type to transform
+ * @return the GraphQL schema string
+ */
+const transformFieldType = (type: IntrospectionTypeRef) => {
+  if (type.kind === 'LIST') {
+    return `[${transformFieldType(type.ofType)}]`;
+  } else if (type.kind === 'NON_NULL') {
+    return `${transformFieldType(type.ofType)}!`;
+  } else if (
+    type.kind === 'SCALAR' ||
+    type.kind === 'OBJECT' ||
+    type.kind === 'ENUM'
+  ) {
+    return type.name;
+  } else {
+    throw new Error(`Unsupported graphql kind ${type.kind} for ${type.name}`);
+  }
+};
+
+/**
+ * Converts an array into an object; for each element of the array, use the prop identified by propName as the key
+ * @param array The array to convert
+ * @param propName The prop name to use as the keys for the new object
+ * @return The object
+ * @throws Error If the array contains more than one element that has the same value for propName
+ * @throws Error If the key propName is not of type string in the array
+ */
+const indexBy = <T>(array: readonly T[], propName: string) => {
+  return array.reduce((acc, curr) => {
+    if (typeof curr[propName] !== 'string') {
+      throw new Error(
+        `key identified by ${propName} in ${curr} must be of type "string"`
+      );
+    }
+    if (curr[propName] in acc) {
+      throw new Error(`key ${curr[propName]} already exists in ${acc}`);
+    }
+    acc[curr[propName]] = curr;
+    return acc;
+  }, {} as { [key: string]: T });
+};
+
+/**
+ * Indexes an a type on an introspection schema using the named type ref, using already indexed types
+ *
+ * @param types The Types that exist on the schema
+ * @param schema The schema to index
+ * @param namedTypeRefKey The named type reference to use for indexing
+ * @param propName The prop name to use as the key
+ * @return the fields of the type identified by thing, indexed using by
+ * @throws Error If the types does not contain a type for the namedTypeRefKey
+ * @throws Error If the array contains more than one element that has the same value for by
+ * @throws Error If the key by is not of type string in the type identified by the namedTypeRefKey
+ */
+const indexSchema = (
```

Comment:
```suggestion
const indexTypeFromSchema = (
```
maybe? naming is hard :)